File: platform/polyglot/github_sdk_api_component_usage_audit/mjs/src/github/endpoints/code-search.mjs
/**
 * GitHub Code Search Endpoint
 *
 * Paginated /search/code with custom iteration for the search API
 * response shape ({ total_count, incomplete_results, items }).
 *
 * Forks are excluded by default. Results capped at 1000 (10 pages x 100).
 */

const PER_PAGE = 100;

/**
 * Search GitHub code for component import patterns.
 *
 * Runs separate queries for each import style (destructured and default)
 * because /search/code doesn't support OR for quoted strings.
 * Results are deduplicated by repo + path across queries.
 *
 * @param {object} ctx - Shared context from the service.
 * @param {Function} ctx.makeSearchRequest - SDK search request function.
 * @param {Function} ctx.log - Logger function.
 * @param {Function} ctx.debugLog - Debug logger function.
 * @param {{ value: boolean }} ctx.cancelled - Cancellation flag.
 * @param {object} params
 * @param {string} params.componentName - Component name to search for.
 * @param {number} params.maxPages - Maximum pages to fetch (1-10).
 * @param {number} params.minFileSize - Minimum file size filter.
 * @param {string} [params.ownerScope] - Optional owner qualifier (e.g. "org:mui" or "user:octocat").
 * @returns {AsyncGenerator<object>} Yields search result items.
 */
export async function* searchCode(ctx, { componentName, maxPages, minFileSize, ownerScope }) {
  const queries = buildSearchQueries(componentName, minFileSize, ownerScope);
  const seen = new Set();

  for (const query of queries) {
    ctx.log(`Searching: ${query}`);
    ctx.log(`Max pages: ${maxPages} (${maxPages * PER_PAGE} results max)`);

    for (let page = 1; page <= maxPages; page++) {
      if (ctx.cancelled.value) {
        ctx.log("Search cancelled", "warn");
        return;
      }

      ctx.log(`Fetching page ${page}/${maxPages}...`);

      const data = await ctx.makeSearchRequest(
        `GET /search/code`,
        {
          q: query,
          per_page: PER_PAGE,
          page,
        },
      );

      await ctx.debugLog("code-search-page", { page, total_count: data.total_count, items: data.items?.length });

      const items = data.items || [];
      if (items.length === 0) {
        ctx.log(`No more results after page ${page - 1}`);
        break;
      }

      for (const item of items) {
        const key = `${item.repository?.full_name}/${item.path}`;
        if (seen.has(key)) continue;
        seen.add(key);
        yield item;
      }

      // Stop if we've fetched all available results
      if (items.length < PER_PAGE) {
        ctx.log(`All results fetched (${page} page${page > 1 ? "s" : ""})`);
        break;
      }
    }
  }
}

/**
 * Build GitHub code search queries for both import styles.
 *
 * GitHub /search/code doesn't support OR for quoted strings, so we
 * return separate queries for destructured and default imports:
 *   - `"import { Typography }"` → `import { Typography } from '...'`
 *   - `"import Typography from"` → `import Typography from '...'`
 *
 * @param {string} componentName - Component name to search for.
 * @param {number} minFileSize - Minimum file size in bytes.
 * @param {string} [ownerScope] - Optional owner qualifier (e.g. "org:mui" or "user:octocat").
 * @returns {string[]} Array of formatted search queries.
 */
export function buildSearchQueries(componentName, minFileSize, ownerScope) {
  const extensions = `extension:tsx extension:jsx size:>${minFileSize}`;
  const scope = ownerScope ? ` ${ownerScope}` : "";

  return [
    `"import { ${componentName} }" ${extensions}${scope}`,
    `"import ${componentName} from" ${extensions}${scope}`,
  ];
}

/** @deprecated Use buildSearchQueries instead. */
export function buildSearchQuery(componentName, minFileSize, ownerScope) {
  return buildSearchQueries(componentName, minFileSize, ownerScope)[0];
}

File: platform/polyglot/github_sdk_api_component_usage_audit/py/github_sdk_api_component_usage_audit/sdk/code_search.py
"""
GitHub Code Search Endpoint

Paginated /search/code with custom iteration for the search API
response shape ({ total_count, incomplete_results, items }).

Forks are excluded by default. Results capped at 1000 (10 pages x 100).
"""

from __future__ import annotations

import logging
from typing import Any, AsyncIterator

from github_api.sdk.client import GitHubClient

__all__ = ["search_code", "build_search_query"]

PER_PAGE = 100

logger = logging.getLogger("github_sdk_api_component_usage_audit.sdk.code_search")


async def search_code(
    client: GitHubClient,
    *,
    component_name: str,
    max_pages: int = 10,
    min_file_size: int = 1000,
) -> AsyncIterator[dict[str, Any]]:
    """Search GitHub code for component import patterns.

    Args:
        client: Authenticated GitHubClient instance.
        component_name: Component name to search for.
        max_pages: Maximum pages to fetch (1-10).
        min_file_size: Minimum file size filter.

    Yields:
        Individual search result items.
    """
    query = build_search_query(component_name, min_file_size)
    logger.info("Searching: %s", query)
    logger.info("Max pages: %d (%d results max)", max_pages, max_pages * PER_PAGE)

    for page in range(1, max_pages + 1):
        logger.info("Fetching page %d/%d...", page, max_pages)

        data = await client.get(
            "/search/code",
            params={
                "q": query,
                "per_page": PER_PAGE,
                "page": page,
            },
        )

        items = data.get("items", [])
        if not items:
            logger.info("No more results after page %d", page - 1)
            return

        for item in items:
            yield item

        # Stop if we've fetched all available results
        if len(items) < PER_PAGE:
            logger.info("All results fetched (%d page%s)", page, "s" if page > 1 else "")
            return


def build_search_query(component_name: str, min_file_size: int) -> str:
    """Build the GitHub code search query string.

    Args:
        component_name: Component name to search for.
        min_file_size: Minimum file size in bytes.

    Returns:
        Formatted search query.
    """
    import_pattern = f'"import {{ {component_name} }}"'
    return f"{import_pattern} extension:tsx extension:jsx size:>{min_file_size}"
